lib/oidcstrategy.js

```diff
@@ -207,350 +299,81 @@ Strategy.prototype.authenticate = function authenticateStrategy(req, options) {
-                      accessToken,
-                      refreshToken,
-                      params,
-                    });
-                  } catch (ex) {
-                    return next(ex);
-                  }
-                });
-              }
-
-              // lets do an id_token fallback. We use id_token over userInfo endpoint for now
-              return onProfileLoaded(self, {
-                req,
-                sub,
-                iss,
-                profile: (idToken) ? makeProfileObject(jwtClaims, jwtClaimsStr) : undefined,
-                jwtClaims,
-                accessToken,
-                refreshToken,
-                params,
-              });
-            });
-          });
-        });
+      /* 
+       * Step 2. load options from metadata url
+       */
+      (metadataUrl, cachekey, next) => {
+        return self.setOptions(self._options, metadataUrl, cachekey, next);
       },
-      //
-      (next) => {
-        if (!(req.body && req.method === 'POST')) {
-          return next();
-        }
 
-        log.info(`OPENID CONNECT: We are in the OpenID Connect Only Response.
-          Assumed because no auth code was in the query and we are POST.`);
-        log.info('Body received was: ', req.body);
-
-        // We are not doing auth code so we set the tokens to null
-        const accessToken = null;
-        const refreshToken = null;
-        const params = null;
-
-        // We have a response, get the user identity out of it
-        const idToken = req.body.id_token;
-        if (!idToken) {
-          return next(new Error('ID Token not present in response'));
+      /* 
+       * Step 3. the following are the scenarios for the coming request
+       * (1) error response
+       * (1) implicit flow (response_type = 'id_token')
+       *     This case we get a 'id_token'
+       * (2) hybrid flow (response_type = 'id_token code')
+       *     This case we get both 'id_token' and 'code'
+       * (3) authorization code flow (response_type = 'code')
+       *     This case we get a 'code', we will use it to get 'access_token' and 'id_token'
+       * (5) for any other request, we will ask for authorization and initialize the authorization process 
+       */
+      (next) => {
+        var err, err_description, id_token, code;
+        err = err_description = id_token = code = null;
+
+        if (req.query && req.query.error) err = req.query.error;
+        if (req.query && req.query.error_description) err_description = req.query.error_description;
+        if (req.query && req.query.id_token) id_token = req.query.id_token;
+        if (req.query && req.query.code) code = req.query.code;
+
+        if (req.body && req.body.error) err = req.body.error;
+        if (req.body && req.body.error_description) err_description = req.body.error_description;
+        if (req.body && req.body.id_token) id_token = req.body.id_token;
+        if (req.body && req.body.code) code = req.body.code;
```

Comment:
We need to make sure the 'id_token' and 'code' came from the POST. this code will allow the 'code' in the query and 'id_token' in the post